### PR TITLE
add syncElementsThatCssClassChanged for [Issue #917]

### PR DESCRIPTION
**Pull request by liuyuanhuo**
update src\mdlComponentHandler.js,add syncElementsThatCssClassChanged for [Issue #917](https://github.com/google/material-design-lite/issues/917), and modify an error in upgradeElementInternal

### CODE REVIEW

**Comment by liuyuanhuo** [comment]
I signed it!

**Comment by Garbee** [comment]
What exactly is this trying to add capability for that isn't currently possible? Or if it can currently be done what is a comparison of code to show before and after effects?

**Comment by liuyuanhuo** [comment]
@Garbee  I add a syncElementsThatCssClassChanged method, this Sync all component for elements that Css Class have changed, it down grade all component for elements that have some cssClass DELETED or ADDED changed, and upgrade elements that have some cssClass ADDED changed. this could solve my all Issue talk at [Issue #917](https://github.com/google/material-design-lite/issues/917),

 could you have a look at the source, do some modify and add it?

then I can use it like below:
```html
 <body aurelia-app="animation-main" onload='setIntervalSyncMDL()'>

    <script>
      function setIntervalSyncMDL()
      {
        setInterval("syncMDL();", 200);
      }
      function syncMDL() {
        componentHandler.syncElementsThatCssClassChanged();
      }
    </script>
```

**Comment by surma** [comment]
Hey @liuyuanhuo. First of all, I appreciate the effort you put in to making this whole thing work with other framworks. However:

Looking at #917, I *strongly* disagree with running a complete downgrade-upgrade-combo every 200ms. That is a performance blackhole. What you are doing in your method should *never* necessary. You are basically encoding all your state in classes, which is not what you should be doing on dynamic sites. Also, you are assuming that elements still exist when you do the downgrade which, when working with certain frameworks, certainly is not guaranteed.

A proper integration with Angular or similar would have to track the state externally and not impose that on MDL. It would call the destructor when the element is being deleted/detached and call the upgrade methods on creation. 

Bottom line: This is not the responsibility of MDL and we are trying to stay true to the “L” in “MDL”.

**Comment by Garbee** [comment]
I agree. From that code sample this gets a big :-1:. There are ways for this to be handled accurately in user-land. We also have contributors exploring new methods (such as in PR #950 ) which will make this kind of thing easier in different situations.

As of right now, we provide every method needed to handle adding new nodes and removing them. *when* they get fired and with what scope (aside from initial upgrade on load), is up to you to decide in user-land.

**Comment by liuyuanhuo** [comment]
I think this is a simple way to use MDL, and can you have a consider on this?
and by the way, there is an error in upgradeElementInternal , and I have add a line to modify it, and have test for it

**Comment by Garbee** [comment]
>  there is an error in upgradeElementInternal , and I have add a line to modify it, and have test for it

What is this exactly? If there is an error in existing code then that warrants its own issue/PR. Let's not go tossing it into this addition.

As far as the merit of the actual addition, it only helps with bad application architecture. Surma gave some very sound logic for why that is so. MDL should not encourage this behavior since it is only sloppy (and with this build very error-prone.)

Since Surma and myself disapprove of this addition, it is highly unlikely to be accepted. I'm going ahead and closing this issue with that in mind. We shouldn't be adding things into core this large to solve some specific problem in architecture someone is having. There is almost always an alternative way.

In the future, please discuss (via an issue) details behind larger code additions like this. It is best to plan things out ahead of time before someone puts time into a large patch. If you need general support, there are communities just for helping developers. StackOverflow for example, there you can use the `material-design-lite` tag among others for angular to try and get help from other developers familiar with the technologies you are using.
